### PR TITLE
tickets/DM-32786: document build process in new sciplat-lab regime

### PR DESCRIPTION
**Pull request by athornton**

### CODE REVIEW

File: index.rst [hunk 1]
```diff
@@ -42,15 +42,183 @@
 
 .. sectnum::
 
-.. TODO: Delete the note below before merging new content to the main branch.
+   This technote documents how sciplat-lab (the JupyterLab RSP container) is built.
```
Comment (jonathansick) [inline index.rst]: This shouldn’t be indented (as it is, it’s being treating as content on `sectnum`, which you don’t want).

File: index.rst [hunk 2]
```diff
@@ -42,15 +42,183 @@
...
 
-.. note::
+.. Add content here.
 
-   **This technote is not yet published.**
+Repository
+==========
 
-   This technote documents how sciplat-lab (the JupyterLab RSP container) is currently built, and outlines some possible future directions to simplify and standardize the process.
+Layout
```
Comment (jonathansick) [inline index.rst:55]: There isn’t a description of the repo layout, so maybe drop this section and instead have “The Lab container resides….” under the Repository header?

Comment (jonathansick) [reply index.rst:55]: Though providing a list of key files in this Layout section could be a good idea.

File: index.rst [hunk 3]
```diff
@@ -42,15 +42,183 @@
...
+------
+
+The Lab container resides at `the LSST SQuaRE Github sciplat-lab
```
Comment (jonathansick) [inline index.rst]:
```suggestion
The Lab container resides at `the LSST SQuaRE GitHub sciplat-lab
```

File: index.rst [hunk 4]
```diff
@@ -42,15 +42,183 @@
...
+Note that from the GitHub perspective, ``prod`` rather than ``main`` is
+the default branch.
+
+Updating the Default Branch
+---------------------------
+
+#. Do your work in a ticket branch, as with any other repository.
+#. PR that ticket branch into ``main``.  Note that the default branch to
+   PR into is going to be ``prod`` and you will have to change the
+   selection to ``main``.
```
Comment (jonathansick) [inline index.rst:99]: This type of workflow will be easy for contributors to mess up. I think it’s important to protect around PRing into the `prod` branch. Here are some ideas:

- Change the default branch to be `main` and then change whatever the CI is to use `prod` as the branch for builds. (this is my recommendation)
- Create a GitHub Action workflow that fails any `pull_request` even onto `prod`

File: index.rst [hunk 5]
```diff
@@ -42,15 +42,183 @@
...
+#. Rebase (if possible) or cherry-pick the changes from ``main`` into
+   ``prod_update``.  At the time of writing, there's no difference
+   between ``main`` and ``prod_update``, but as we migrate between major
+   versions of JupyterLab, it is possible for the two branches to
+   diverge significantly (as they did in the JL2-JL3 transition).
```
Comment (jonathansick) [inline index.rst:104]: So during a major transition, `main` would preview the next JL version, and `prod_update` would interface changes from the next JL to the old JL build from `prod`?

File: index.rst [hunk 6]
```diff
@@ -42,15 +42,183 @@
...
+If the image is built from a branch that is not ``prod``, and the
+``supplementary`` tag is not specified, the supplementary tag will be
+set to a value derived from the branch name.  This prevents building
+standard containers from branches other than ``prod``.
+
+Dockerfile template substitution
+--------------------------------
+`Dockerfile.template
+<https://github.com/lsst-sqre/sciplat-lab/blob/main/Dockerfile.template>`_
+looks like it's ready for Jinja 2: we're substituting ``{{TAG}}``,
```
Comment (jonathansick) [inline index.rst]: Maybe rephrase? I’m a bit confused by “looks like it’s ready for Jinja 2”

File: index.rst [hunk 7]
```diff
@@ -42,15 +42,183 @@
...
+``{{IMAGE}}``, and ``{{VERSION}}``.  It's nothing that sophisticated.
+We just run ``sed`` in the ``dockerfile`` target of the `Makefile
+<https://github.com/lsst-sqre/sciplat-lab/blob/main/Makefile>`_.
+
+
+Examples
+--------
+
+Build and push the weekly 2021_50 container: ``make tag=w_2021_50``.
```
Comment (jonathansick) [inline index.rst]: I think it’ll be easier to read these if they’re introductory sentences with a code-block that follows, like this:

```suggestion
Build and push the weekly 2021_50 container::

    make tag=w_2021_50
```

File: index.rst [hunk 8]
```diff
@@ -42,15 +42,183 @@
...
+
+stage*.sh scripts
+-----------------
+
+Most of the action in the Dockerfile comes from five shell scripts
+executed by ``docker build`` as ``RUN`` actions.
+
+These are, in order:
+
+#. rpm -- we will always be building on top of ``centos`` in the current
```
Comment (jonathansick) [inline index.rst]: I’d recommend spelling out the whole file name (e.g. `stage1-rpm.sh`). This is so that it’s easier to search.

**Review by jonathansick — APPROVED** [review]
After merging this, make sure to update the README of the sciplat-lab repo so that links to this technote as its “documentation.”

File: index.rst (hunk 5 above)
Comment (athornton) [reply index.rst:104]: Correct.

File: index.rst (hunk 1 above)
Comment (athornton) [reply index.rst]: Fixed.

File: index.rst (hunk 3 above)
Comment (athornton) [reply index.rst]: Fixed.

File: index.rst (hunk 6 above)
Comment (athornton) [reply index.rst]: Done.

File: index.rst (hunk 7 above)
Comment (athornton) [reply index.rst]: Done.

File: index.rst (hunk 4 above)
Comment (athornton) [reply index.rst:99]: That gets tricky in that the workflow, as I understand it, has to run from the default branch.  It is a tricky workflow, though.  I'm going to punt this for now, but your point is well taken.

File: index.rst (hunk 8 above)
Comment (athornton) [reply index.rst]: Done.

File: index.rst (hunk 2 above)
Comment (athornton) [reply index.rst:55]: Done.
